sammba/modality_processors/perfusion_fair.py

```diff
@@ -0,0 +1,369 @@
+
+
+#the perfusion fluid-attenuated inversion-recovery function
+#the jacobian should be callable but I do not know enough maths to create it
+def _fair_t1_func(x, ti, s0):
+    return x[0] + np.absolute(x[1] * (1.0 - 2.0 * np.exp(-ti / x[2]))) - s0
+
+
+def fair_t1_fit(s0, ti, t1_guess):
+    """
+    Fit the perfusion FAIR equation:
+        
+        signal = bias + abs(M0 * (1 - 2 * (exp(-TI/T1))))
+        
+    using scipy.optimize.least_squares (default Trust Region Reflective
+    algorithm). Bias, M0 and T1 are all estimated with a zero lower bound.
+    Starting values for bias (zero) and M0 (mean of input signals) are
+    calculated automatically; T1 has to be supplied, with the mean of TIs
+    being a good guess.
+    
+    Parameters
+    ----------
+    s0 : numpy array of int or float
+        The aquired signals.
+        
+    ti : list of int or float
+        The inversion times. Must have the same length as s0.
+        
+    t1_guess : int or float
+        An initial starting value for T1. The mean of TIs is a good guess.
+
+    Returns
+    -------
+    class scipy.optimize.OptimizeResult    
+    
+    Notes
+    ----- 
+    Levenberg-Marquardt (method='lm'), the usual literature algorithm, cannot
+    be used here as the implementation in scipy does not accept bounds. From
+    experience with the R MINPACK implementation (minpack.lm function nlsLM),
+    these bounds are necessary to assure good fitting. The lmfit python package
+    does implement bounds for MINPACK Levenberg-Marquardt (method=leastsq;
+    default), though the results appear to be worse than for TRR
+    (method=least_squares), whose results bizarrely look closer to those of R
+    minpack.lm nlsLM (to confirm), and are of course identical to
+    scipy.optimize.least_squares with method='trf' (default).
+    """
+    
+    return ls(_fair_t1_func, np.array([0, np.mean(s0), t1_guess]),
+              bounds=([0, 0, 0], np.inf), args=(np.array(ti), s0))
+
+
+def perf_fair_fitter(s0, t1_blood, lambda_blood, ti, multiplier, t1_guess,
+                     picker_sel, picker_nonsel, outtype='simple'):
+    """
+    Wrapper to execute fair_t1_fit on a real signal vector containing both
+    selective and non-selective inversions. Also calculates rCBF and absolute
+    CBF from the resulting selective and non-selective T1 values.
+    
+    Parameters
+    ----------
+    s0 : numpy array of int or float
+        The aquired signals.
+        
+    t1_blood : int or float
+        T1 of blood in ms at the acquisition field strength. Empirical
+        constant; an example value is 2800 at 11.7T (Duong ref).
+        
+    lambda_blood : int or float
+        The assumed blood tissue partition coefficient of water in ml per g.
+        Empirical constant; usually 0.9.
+        
+    ti : list of int or float
+        The inversion times. Must have the same length as s0.
+        
+    multiplier : int or float

```

Comment:
may be this parameter can be removed and tell instead that you are returning CBF in ml per g per ms. Users interested in another limit can do the conversion themselves.

Reply:
It's a literature convention, the best that can be done is to provide default values